native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/avro/AvroSerializer.cs

```diff
@@ -0,0 +1,135 @@
+    {
+
+        private readonly MemoryCache _datumWriterCache;
+
+        /// <summary>
+        /// Size of datum writer cache for testing purpose
+        /// </summary>
+        public int CacheSize => _datumWriterCache.Count;
+
+        private AvroSerializer()
+        {
+            // TODO: make cache size limit configurable
+            _datumWriterCache = new MemoryCache(new MemoryCacheOptions{SizeLimit = 1000});
+        }
+
+        /// <summary>
+        ///  Constructor for Avro serializer that accepts configuration elements.
+        /// </summary>
+        /// <param name="configuration">configuration elements</param>
+        public AvroSerializer(GlueSchemaRegistryConfiguration configuration = null)
+        :this()
+        {
+        }
+
+        /// <summary>
+        /// Serialize the Avro message to bytes
+        /// </summary>
+        /// <param name="data">the Avro message for serialization</param>
+        /// <returns>the serialized byte array</returns>
+        public byte[] Serialize(object data)
+        {
+            // ReSharper disable once ConvertIfStatementToReturnStatement
+            if (data == null)
+                return null;
+            else
+                return Serialize(data, GetOrCreateDatumWriter(data));
+        }
+
+        /// <summary>
+        /// Get the schema definition.
+        /// </summary>
+        /// <param name="data">object for which schema definition has to be derived</param>
+        /// <returns>schema string</returns>
+        public string GetSchemaDefinition(object data)
+        {
+            var schema = GetSchema(data);
+            return schema.ToString();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="schemaDefinition"></param>
+        /// <param name="data"></param>
+        public void Validate(string schemaDefinition, byte[] data)
+        {
+            //No-op
+            //We cannot determine accurately if the data bytes match the schema as Avro bytes don't contain the field names.
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="data"></param>
+        public void Validate(object data)
+        {
+            //No-op
+            //Avro format assumes that the passed object contains schema and data that are mutually conformant.
+            //We cannot validate the data against the schema.
+        }
+
+        private static byte[] Serialize(object data, DatumWriter<object> datumWriter)
+        {
+            var memoryStream = new MemoryStream();
+            Encoder encoder = new BinaryEncoder(memoryStream);
+            datumWriter.Write(data, encoder);
+            return memoryStream.ToArray();
+        }
+
+        private DatumWriter<object> GetOrCreateDatumWriter(object data)
+        {
+            var schema = GetSchema(data);
+            var dataType = data.GetType().ToString();
+            var cacheKey = schema + dataType;
```

Comment:
What's the data type of cacheKey?

Reply:
string. I am appending dataType to schema definition in order to differentiate specific record and generic record with same schema def